File: ProblemSolving/ProblemSolving/Programmers/Level3/PG42627_디스크컨트롤러.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

struct INFO {
    int s, w;
};

priority_queue<INFO> pq;

bool operator<(INFO a, INFO b) {
    if (a.w == b.w) {
        return a.s > b.s;
    }
    return a.w > b.w;
}

int solution(vector<vector<int>> jobs) {
    int answer = 0;

    sort(jobs.begin(), jobs.end());

    int t = 0;
    int index = 0;
    int jSize = jobs.size();
    while (1) {
        if (index == jSize && pq.empty())
            break;

        // 현재 시간을 기준으로 작업에 들어갈 수 있는 대상을 힙에 넣기
        for (index; index < jSize; index++) {
            int s = jobs[index][0];
            int w = jobs[index][1];

            if (s > t)
                break;

            pq.push({ s, w });
        }

        if (pq.empty()) {
            if(index < jSize)
                t = jobs[index][0];
            continue;
        }

        INFO target = pq.top();
        pq.pop();
        t += target.w;
        answer += (t - target.s);
    }

    return answer/jSize;
}

int main() {

    cout << solution({ {0,3},{1,9},{2,6} });

    return 0;
}

File: ProblemSolving_Java/ProblemSolving_Java/src/programmers/level3/PG42627_디스크컨트롤러.java
package programmers.level3;

import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

public class PG42627_디스크컨트롤러 {

    public static void main(String[] args) {
        int[][] jobs = {{0, 3}, {2, 6}, {1, 9}};
        Solution solution = new Solution();
        int answer = solution.solution(jobs);
        System.out.println("answer = " + answer);
    }

    static class Solution {

        PriorityQueue<int[]> pq = new PriorityQueue<>(new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                if (o1[1] == o2[1]) {
                    return o1[0] - o2[0];
                }
                return o1[1] - o2[1];
            }
        });

        public int solution(int[][] jobs) {


            // 요청 시간, 작업 시간을 기준으로 오름차순 정렬하기
            Arrays.sort(jobs, new Comparator<int[]>() {
                @Override
                public int compare(int[] o1, int[] o2) {
                    if (o1[0] == o2[0]) {
                        return o1[1] - o2[1];
                    }
                    return o1[0] - o2[0];
                }
            });

            int t = 0;      // 현재시간
            int index = 0;  // jobs index
            int sum = 0;    // 요청~종료 누적시간
            while (true) {
                // 현재 시간을 기준으로 실행시킬 수 있는 작업들을 힙에 넣기
                for (; index < jobs.length; index++) {
                    if (jobs[index][0] <= t)
                        pq.add(jobs[index]);
                    else
                        break;
                }

                // 만약 pq가 비어있다면, 현재 시간을 가장 먼저 온 요청 시간으로 세팅
                if (pq.isEmpty()) {

                    // 힙에 넣을 작업이 더 이상 존재하지 않는 경우
                    if (index == jobs.length) break;

                    pq.add(jobs[index]);
                    t = jobs[index][0];
                    index++;
                }

                // pq에서 top을 하나 꺼냄
                // 힙은 작업시간을 기준으로 제일 짧은 작업을 선택
                int[] top = pq.poll();

                sum += (t + top[1] - top[0]);   // 대기시간 누적
                t = t + top[1];               // 현재시간 변경
            }

            return sum / jobs.length;
        }

    }
}
